text/0000-ppv.md

```diff
@@ -0,0 +1,1471 @@
+vectors are "small" by most platforms standards.
+
+These types are useful for performing Simd Within A Register (SWAR) operations
+in platforms without SIMD registers. While their performance has not been
+extensively investigated in `stdsimd` yet, any performance issues are
+performance bugs that should be fixed.
+
+### Portable shuffles API
+
+The portable shuffles are exposed via the `shuffle!` macro. Generating the
+sequence of instructions required to perform a shuffle requires the shuffle
+indices to be known at compile time.
+
+In the future, an alternative API based on `const`-generics and/or
+`const`-function-arguments could be added in a backwards compatible way:
+
+```rust
+impl {element_type}{element_width}x{number_of_lanes} {
+    pub fn shuffle<const N: usize, R>(self, const indices: [isize; N]) 
+        -> <R as ShuffleResult<element_type, [isize; N]>>::ShuffleResultType
+      where R: ShuffleResult<element_type, [isize; N]>;
+}
+```
+
+Offering this same API today is doable:
+
+```rust
+impl {element_type}{element_width}x{number_of_lanes} {
+    #[rustc_const_argument(2)] // specifies that indices must be a const
+    #[rustc_platform_intrinsic(simd_shuffle2)]
+    // ^^^ specifies that this method should be treated as the 
+    // "platform-intrinsic" "simd_shuffle1"
+    pub fn shuffle2<I>(self, other: Self, indices: I) 
+        -> <R as ShuffleResult<element_type, I>>::ShuffleResultType
+      where R: ShuffleResult<element_type, I>;
+      
+    #[rustc_const_argument(1)]
+    #[rustc_platform_intrinsic(simd_shuffle1)]
+    pub fn shuffle<I>(self, indices: I) 
+        -> <R as ShuffleResult<element_type, I>>::ShuffleResultType
+      where R: ShuffleResult<element_type, I>;
+}
+```
+
+If there is consensus for it the RFC can be easily amended. 
+
+# Prior art
+[prior-art]: #prior-art
+
+Most of this is implemented in `stdsimd` and can be used on nightly today via
+the `std::simd` module. The `stdsimd` crate is an effort started by @burntsushi
+to put the `rust-lang-nursery/simd` crate into a state suitable for
+stabilization. The `rust-lang-nursery/simd` crate was mainly developed by @huonw
+and IIRC it is heavily-inspired by Dart's SIMD which is from where the `f32x4`
+naming scheme comes from. This RFC has been heavily inspired by Dart, and two of
+the three examples used in the motivation come from the [Using SIMD in
+Dart](https://www.dartlang.org/articles/dart-vm/simd) article written by John
+McCutchan. Some of the key ideas of this RFC come from LLVM's design, which was
+originally inspired by GCC's vector extensions, which was probably inspired by
+something else.
+
+Or in other words: to the author's best knowledge, this RFC does not contain any
+really novel ideas. Instead, it only draws inspriation from previous designs
+that have withstood the test of time, and it adapts these designs to Rust.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+### Interaction with scalable vectors
```

Comment:
Nit but I think "scalable" is very much an Arm-ism here. The more neutral and common term is probably "Cray-style".

Reply:
Yes I agree. In the introduction I mention:

> packed: means that these vectors have a compile-time fixed size. It is the opposite of scalable or "Cray vectors", which are SIMD vector types with a dynamic size, that is, whose size is only known at run-time.

and I consistently used the term "scalable" throughout the RFC because I thought "Cray vectors" would be a less approachable term. I don't like the term "scalable" much either TBH.